Add basic TD framework

ZainRizvi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107243
* #107070
* __->__ #106997

Adds a new structure to house all heuristics we use for Target Determination and Test Reordering.  I'm keeping it somewhat minimal for now, to let it evolve more easily as we try new things.

It currently does nothing. The 2nd pr in the stack ports the existing heuristics to actually use this new framework

ZainRizvi committed:
Add basic TD framework

[ghstack-poisoned]

ZainRizvi committed:
Update on "Add basic TD framework"


Adds a new structure to house all heuristics we use for Target Determination and Test Reordering.  I'm keeping it somewhat minimal for now, to let it evolve more easily as we try new things.

It currently does nothing. The 2nd pr in the stack ports the existing heuristics to actually use this new framework

[ghstack-poisoned]

clee2000 commented on tools/testing/target_determination/heuristics/interface.py:
```diff
@@ -0,0 +1,86 @@
+from abc import abstractmethod
```
typo?

ZainRizvi replied:
Surprisingly, that is the legit package name. Stands for "Abstract Base Class"

clee2000 commented on tools/testing/target_determination/heuristics/interface.py:
```diff
@@ -0,0 +1,86 @@
+    probably_relevant: List[str]
+    unranked_relevance: List[
+        str
+    ]  # When we don't know if the test is relevant to the PR or not
```
any chance you could get the autoformatter to not make not look weird lol

ZainRizvi replied:
👹

clee2000 commented on tools/testing/target_determination/heuristics/interface.py:
```diff
@@ -0,0 +1,86 @@
+
+        self.highly_relevant = TestPrioritizations._merge_tests(
+            self.highly_relevant, other.highly_relevant, exclude_tests=[]
+        )
```
why is it necessary to merge highly relevant tests after extending them?

clee2000 commented on tools/testing/target_determination/determinator.py:
```diff
@@ -0,0 +1,64 @@
+    if num_tests_analyzed != len(tests):
+        warn(
+            f"Was given {len(tests)} tests to prioritize, but only analyzed {num_tests_analyzed} tests"
+        )
```
in what case would you expect this to be not equal?  I would have thought that it should be an assertion, not just a warning

ZainRizvi replied:
Never. Switching it to an assert

clee2000 commented on tools/testing/target_determination/heuristics/interface.py:
```diff
@@ -0,0 +1,86 @@
+        ranked_tests = self.highly_relevant + self.probably_relevant
+        self.unranked_relevance = [
+            t for t in self.unranked_relevance if t not in ranked_tests
+        ]
```
I think it might be worth adding an assertion somewhere that total tests at beginning = total tests at the end

ZainRizvi commented on tools/testing/target_determination/determinator.py:
```diff
@@ -0,0 +1,64 @@
+    emit_metric(
+        "test_reordering_prioritized_tests",
+        {
+            "highly_relevant_test_count": len(rankings.highly_relevant),
```
I do suspect we'll want to shift this to emit separate metrics for each heuristic soon. Wanted to keep the functional changes minimal for this PR though.

And you're right that length is prob not needed. Dropping it

ZainRizvi committed:
Update on "Add basic TD framework"


Adds a new structure to house all heuristics we use for Target Determination and Test Reordering.  I'm keeping it somewhat minimal for now, to let it evolve more easily as we try new things.

It currently does nothing. The 2nd pr in the stack ports the existing heuristics to actually use this new framework

[ghstack-poisoned]

ZainRizvi committed:
Update on "Add basic TD framework"


Adds a new structure to house all heuristics we use for Target Determination and Test Reordering.  I'm keeping it somewhat minimal for now, to let it evolve more easily as we try new things.

It currently does nothing. The 2nd pr in the stack ports the existing heuristics to actually use this new framework

[ghstack-poisoned]

huydhn commented on tools/testing/target_determination/determinator.py:
```diff
@@ -0,0 +1,58 @@
+        + len(rankings.probably_relevant)
+        + len(rankings.unranked_relevance)
+    )
+    assert num_tests_analyzed != len(
```
Could you briefly explain what is the goal of this assert?  I would naively expect `num_tests_analyzed` to be equal to the length of all input tests, but the assert says otherwise.

ZainRizvi replied:
You're right. This was a buggy assert and I just realized I'd accidentally put the fix for this [in a PR](https://github.com/pytorch/pytorch/pull/107071/files#diff-167301b680ef3cd2de2c553db17e6a6c9e46936d3beb1e689149691cbca3b93dL46) higher up in the stack

ZainRizvi committed:
Update on "Add basic TD framework"


Adds a new structure to house all heuristics we use for Target Determination and Test Reordering.  I'm keeping it somewhat minimal for now, to let it evolve more easily as we try new things.

It currently does nothing. The 2nd pr in the stack ports the existing heuristics to actually use this new framework

[ghstack-poisoned]

ZainRizvi commented:
@pytorchbot merge
